ogr/services/pagure.py

```diff
@@ -24,113 +27,224 @@ def __init__(
+
+        self.insecure = insecure
+        if self.insecure:
+            self.session.mount("http://", adapter)
+        else:
+            self.session.mount("https://", adapter)
+
+        if self._token:
+            self.header = {"Authorization": "token " + self._token}
+
     def get_project(self, **kwargs) -> "PagureProject":
-        project_kwargs = self.pagure_kwargs.copy()
-        project_kwargs.update(kwargs)
-        return PagureProject(
-            instance_url=self.instance_url,
-            token=self._token,
-            service=self,
-            read_only=self.read_only,
-            **project_kwargs,
-        )
+        return PagureProject(service=self, **kwargs)
 
     @property
     def user(self) -> "PagureUser":
         return PagureUser(service=self)
 
-    def change_token(self, new_token: str) -> None:
+    def call_api(
+        self,
+        url: str,
+        method: str = None,
+        params: dict = None,
+        data=None,
+        raw: bool = False,
+    ) -> Union[dict, requests.Response]:
+        """ Method used to call the API.
+        It returns the raw JSON returned by the API or raises an exception
+        if something goes wrong.
+
+        It can return raw response when raw=True.
         """
-        Change an API token.
+        method = method or "GET"
+
+        try:
+            response = self.session.request(
+                method=method,
+                url=url,
+                params=params,
+                data=data,
+                verify=not self.insecure,
+                headers=self.header,
+            )
+        except requests.exceptions.ConnectionError as er:
+            logger.error(er)
+            raise PagureAPIException(f"Cannot connect to url: `{url}`.", er)
+
+        if raw:
+            return response
+
+        if response.status_code == 404:
+            raise PagureAPIException(f"Page `{url}` not found when calling Pagure API.")
+
+        try:
+            output = response.json()
+        except ValueError as err:
+            logger.debug(response.text)
+            raise PagureAPIException("Error while decoding JSON: {0}".format(err))
+
+        if not response.ok:
+            logger.error(output)
+            if "error" in output:
+                error_msg = output["error"]
+                raise PagureAPIException(
+                    f"Pagure API returned an error when calling `{url}`: {error_msg}",
+                    pagure_error=error_msg,
+                )
+            raise PagureAPIException(f"Problem with Pagure API when calling `{url}`")
+
+        return output
+
+    def get_raw_request(self, url, method="GET", params=None, data=None):
+
+        req = self.session.request(
+            method=method,
+            url=url,
+            params=params,
+            headers=self.header,
+            data=data,
+            verify=not self.insecure,
+        )
+        return req
```

Comment:
please inspire this code by: https://github.com/PyGithub/PyGithub/blob/master/github/Requester.py#L367
especially return of this method: ``return status, responseHeaders, output`` what then could be easily used for mocking